[9.0] mis_builder Sorrento

sbidoul opened the pull request:
This is the work done on mis_builder at the Sorrento Sprint.

Thanks to all who contributed code and ideas! It was a great event.

This branch should be usable as is, although not yet perfect. 

Sprinters (in particular @adrienpeiffer @sebastienbeau @gfcapalbo), feel free to commit your last small changes here if any. 

Then let's review and merge fast so the work in account_financial_report_qweb that depends on this can continue.

sbidoul committed:
Migrate aep to 9.0: periods are gone, so far, so good

sbidoul committed:
add v9 thoughts in the readme

sbidoul committed:
no more parent_id, no more consolidation children

Code is simpler too but here we really lose functionality.

Commit:
no more fiscal period, no more root account

Commit:
Not really the same js in 9.0 ...

Commit:
[mis_builder] First installable version in 9.0 ! Enjoy !

sbidoul committed:
Bump version number

sbidoul committed:
[FIX] remove unused import

Commit:
[CHG][mis_builder] Use base_report_xlsx instead of report_xls

Commit:
[FIX][mis_builder] Change parent menu for MIS report template to be compatible with Odoo enterprise edition.

Commit:
[IMP][mis_builder] Remove unnecessary code

Commit:
[FIX][mis_builder] Fix test suite.

sbidoul committed:
pep8 & co

Commit:
[IMP][mis_builder] MIS Builder widget improvement, add print and export button in the widget.

Commit:
[IMP][mis_builder] use get_context method in drilldown method

Commit:
[IMP][mis_builder] Improve buttons position

Commit:
[FIX] Flake8 issues

Commit:
[FIX] Bad modification

JordiBForgeFlow committed:
[IMP] add prefixes. comes from PR #145

Commit:
mis_builder : in compute do not set val in localdict if there was a computation error

ThomasBinsfeld committed:
[ADD] AccountingNone (singleton) to differentiate balances among which the debit and the credit are zero and balances among which debit and credit nullify

sbidoul committed:
mis_builder: better layout of mis report screen on small screens

Commit:
[IMP][mis_builder] Improve usability : Display directly report instance instead of report instance settings.

Commit:
[IMP][mis_builder] Add whitespace between buttons and table.

sbidoul committed:
[FIX] reset permissions that should not have changed

sbidoul committed:
[IMP] improve AccountingNone wrt comparisons mainly

sbidoul committed:
Add copyright header and __all__ to accounting_none.py

sbidoul committed:
[FIX] fix comparison rendering in presence of AccountingNone

Commit:
[IMP][mis_builder] Check if current user is in accountant group to show setings button

Commit:
[FIX][mis_builder] Fix the creation of a new mis.report.instance.

sbidoul committed:
mis_builder: render blank instead of +0 comparison columns

sbidoul committed:
mis_builder: better form layout for MIS Report templates

sbidoul committed:
[IMP] mis_builder: refactor the compute method

The main computation entry point is now on mis.report instead of
mis.report.instance.period so it can easily be reused by other
module who need to evaluate mis.report's on without creating
full-fledge persistant mis.report.instance (such as a wizard
that allows the user to select the period for a one-off run).

sbidoul committed:
[IMP] mis_builder: meaningful copy for MIS Report templates

Commit:
[CHG][mis_builder] According with the rename of base_report_xlsx to report_xlsx.

sbidoul committed:
[FIX] mis_builder: fix uninitialized variable

sbidoul committed:
[IMP] mis_builder: copy semantics for mis.report.instance

Commit:
[FIX] Fix flake8 issues

sbidoul committed:
[ADD] mis_builder: simple_array

sbidoul committed:
[FIX] some colspan in mis_builder

Commit:
[ADD][mis_builder] Add the possibility to select date range in mis report period.

sbidoul committed:
[WIP] domain for initial balance depending on account type

Commit:
[FIX][mis_builder] Better with company

Commit:
[CHG][mis_builder] Use company to get date_from

Commit:
[FIX] Flake8 issues

Commit:
[FIX][mis_builder] Don't forget the company for sone_parsing method.

sbidoul committed:
[FIX] mis_builder: fix date type for fiscal year date start computation

plus a bit of pep 8 and idiom to get ids from a recordset

Commit:
[IMP] add a form view for kpi report line

Commit:
[DEL] delete field sequence on the form kpi

sbidoul committed:
mis_builder: add changelog

sbidoul committed:
[FIX] mis_builder: use =like instead of like to search on account codes

sebastienbeau committed:
[IMP] start adding the sub kpi concept

sebastienbeau committed:
[IMP] improve view and onchange on multi expression

sebastienbeau committed:
[FIX] fix operation on SimpleArray

sebastienbeau committed:
[FIX] fix ordering

sebastienbeau committed:
[REF] remove dead code

sebastienbeau committed:
[FIX] fix sum of simple value

sebastienbeau committed:
[FIX] do not insert in localdict if there is error

sebastienbeau committed:
[FIX] fix error with simple expression

sbidoul committed:
[IMP] mis_builder: some doc, minor refactoring and pep8 stuff around subkpis

lmignon committed:
[DEL] Remove xls report (replaced by xlsx)

lmignon committed:
[FIX] correctly handle Sub KPI in xlsx report

lmignon committed:
[FIX] correctly handle Sub KPI in qweb report

sbidoul committed:
[WIP] mis_builder auto-detail

sbidoul committed:
[WIP] mis_builder: disable the WIP auto-detail feature

sebastienbeau committed:
[IMP] improve report view

sbidoul committed:
[WIP] mis_builder: auto-expand by account feature

lmignon committed:
[IMP] Refactor code and implement styling in xlsx

lmignon committed:
[IMP] Add Giovanni Capalbo in the list of authors

sbidoul committed:
[WIP] mis_builder: display account code/names in auto-expanded details

sbidoul committed:
cosmetics

sebastienbeau committed:
[IMP] add possibility to use a report as a wizard

sebastienbeau committed:
[IMP] description on report instance is now a related to the description done on the report

gfcapalbo committed:
[ADD] New MIS-Builder Styles

sbidoul committed:
[FIX] mis_builder: fix comparison_mode compute

Commit:
[FIX][mis_builder] Duplicate xml_id in the same module.

sbidoul committed:
[FIX] mis_builder test back to green

sbidoul committed:
flake8

sbidoul committed:
[FIX] add missing mis_builder acls

sbidoul committed:
[FIX] wrong group in acl

lmignon committed:
[FIX] Add necessary dependencies for date_range used by mis_builder

Commit:
[IMP][mis_builder] Add default style for mis_builder report.

sebastienbeau committed:
[IMP] add cron for deleting temporary report, improve view and fix period creation

sebastienbeau committed:
[FIX] fix default value at creation of the computed field comparison_mode

Commit:
[FIX] Display date_from and date_to with date_range.

Commit:
[IMP] Improve style of Mis report widget buttons. Add report name in the widget.

sbidoul committed:
[FIX] mis_builder: fixes wrt subkpis and drilldown

sbidoul committed:
[IMP] mis_builder changlog

sbidoul committed:
[IMP] mis_builder: add doctests to the test suite

sbidoul committed:
[IMP] mis_builder: add 'u' mode for unallocated p&l

sbidoul committed:
[IMP] mis_builder: split computation of ending balance in initial+variation

So if there are no moves in the period, the ending balance is
AccountingNone instead of 0.

sbidoul committed:
[IMP] mis_builder: refine handling of null initial balances

Filter out accounts which have a 0 contribution to the initial balance

sbidoul committed:
[IMP] mis_builder: support SimpleArray operations with ints/floats

sbidoul committed:
tweak readme

sbidoul committed:
[FIX] mis_builder: unallocated p&l computation fix

sbidoul committed:
[IMP] mis_builder: unit tests for AccountingExpressionProcessor

sbidoul committed:
pep8

sbidoul committed:
a couple more tests

sbidoul committed:
[IMP] mis_builder: give a description to sub kpis

name is the python variable name, to be used (in the future)
to reference sub-kpi values in expressions.

description is the human-readable name to display in column
headers.

sbidoul committed:
[WIP] mis_builder: refactoring

sbidoul committed:
[IMP] mis_builder: convenience methods to obtain balance and unallocated p&l

plus improve mechanism to group by account

sbidoul committed:
[IMP] mis_builder: increase aep test coverage

sbidoul committed:
[IMP] mis_builder improve AEP convenience methods

Commit:
[IMP][mis_builder] Add print, export and preview buttons directly on mis.report.instance tree view.

sbidoul commented:
A layout issue in the pdf export wrt sbukpis: 
![image](https://cloud.githubusercontent.com/assets/692075/14954180/911784f6-1070-11e6-9a34-c28eb37b38a6.png)

sbidoul commented:
https://github.com/OCA/account-financial-reporting/pull/189/commits/6c213092d48e98b06a80fce8d83f123c9c1276a9 is nice thanks! Is it possible to disable the Edit button on the preview form?

adrienpeiffer commented:
@sbidoul Yes. And the create button too ?

sbidoul commented:
@adrienpeiffer currently the create button on the preview form has a weird behaviour so I'd hide it too. Delete and duplicate do not make much sense either, BTW.

adrienpeiffer commented:
That's what I thought. I do this ASAP

Commit:
[IMP][mis_builder] Avoid to create, edit, delete on mis.report.instance widget form.

adrienpeiffer commented:
@sbidoul it's done !

jgrandguillaume commented:
Hi,

I've just spend some times testing this PR. All in all seems good enough to be merged IMO. Still some work to be done to have something nice and user friendly, but it is just a huge step forward already.

I'm :+1:  CI green and once on the official repo, people will start improving it more than if we wait too long before merging this one.

Thanks a lot for the work !

Regards,

Joël

sbidoul committed:
[WIP] mis_builder: refactoring

sbidoul committed:
[IMP] make SimpleArray produce DataError in memberwise operations

Put DataError and mis_safe_eval in their own files

sbidoul committed:
[IMP] Move the MIS Report Templates to the Accounting > Config > Financial Reports menu

sbidoul committed:
[IMP] mis_builder: AEP: replace_exprs_by_account_id now works on a list of expressions

sbidoul committed:
[IMP] remove unused doctests call

sbidoul committed:
[IMP] mis_builder: tests for mis_safe_eval

sbidoul committed:
[WIP] mis_builder refactoring: cette fois je tiens le bon bout :)

sbidoul committed:
[FIX] mis_builder: aep issue with smart ending balance computation

sbidoul committed:
[WIP] mis_builder refactoring: restore proper display and sorting of detail rows

sbidoul committed:
[WIP] mis_builder refactoring: restore period titles in widget

sbidoul committed:
[WIP] mis_builder refactoring: fix bug with detail accounts

sbidoul committed:
[FIX] mis_builder: widget footer was broken

sbidoul committed:
[FIX] mis_builder: fix period comment (date from/to) display

sbidoul committed:
[WIP] mis_builder refactoring: remove old, ugly code \o/

sbidoul commented:
While looking at adapting the code to the new features (comparison, drilldown), it became clear that mis_builder.py is showing signs of age. I'm in the middle of a major refactoring that will make it much more maintainable. Should be done in a few days. Please wait before merging.

sbidoul committed:
[WIP] mis_builder refactoring: restore and improve comparison columns

sbidoul committed:
[WIP] mis_builder refactoring: move the "json" conversion to the matrix

mis.report.instance.compute() is now as simple as possible, and rest
is nicely factored in manageable methods \o/

sbidoul committed:
[WIP] mis_builder refactoring: split mis_builder.py in two

mis_report.py has the heavy stuff, and contains all the computation
and rendering code

sbidoul committed:
[FIX] mis_builder: a few bug in mis_builder_style

sbidoul committed:
[WIP] mis_builder refactoring: restore tooltips and make them work with subkpis

sbidoul committed:
[FIX] mis_builder: remove redundant css

sbidoul committed:
[WIP] mis_builder refactoring: split compute method

So the matrix computation can be used without the json rendering.

sbidoul committed:
[WIP] mis_builder refactoring: cleanup widget styles a bit

sbidoul committed:
[WIP] mis_builder refactoring: refactor and improve PDF report

sbidoul committed:
[FIX] mis_builder: minor tweak to avoid empty style attributs in widget

sbidoul committed:
[WIP] mis_builder refactoring: restore and improve xslx

Including decent column widths.

sbidoul committed:
[FIX] mis_builder: correct use of color picker widget

sbidoul committed:
[WIP] mis_builder refactoring: fix and clean the test

sbidoul committed:
[IMP] mis_builder: style for auto-expanded detail rows

sbidoul committed:
[IMP] mis_builder: add no-break space for better display of multi-formula expressions

sbidoul committed:
[FIX] mis_builder: do not use color picker widget because it does not allow for null values

sbidoul committed:
[IMP] mis_builder: cosmetics in style form

sbidoul committed:
[IMP] mis_builder: update contributors in README

sbidoul committed:
[WIP] mis_builder refactoring: rename view and report sources for clarity

sbidoul committed:
[FIX] mis_builder: xlsx: fix header display issue with simple kpis

sbidoul committed:
[WIP] mis_builder refactoring: add a menu entry for styles

Plus do some more source renaming around styles.

sbidoul committed:
[IMP] mis_builder: changelog

sbidoul committed:
[LNT] mis_builder: yes, I do want to use eval()

sbidoul commented on mis_builder/models/mis_safe_eval.py:
```diff
@@ -0,0 +1,33 @@
+    try:
+        c = test_expr(expr, _SAFE_OPCODES, mode='eval')
+        globals_dict = {'__builtins__': _BUILTINS}
+        val = eval(c, globals_dict, locals_dict)  # noqa
```
@moylop260 how can I silence `W0123(eval-used)`? I really need to use `eval()` here.

moylop260 replied:
``` diff
       globals_dict = {'__builtins__': _BUILTINS}
-      val = eval(c, globals_dict, locals_dict)  # noqa
+      val = eval(c, globals_dict, locals_dict)  # pylint: disable=eval-used
```

sbidoul replied:
@moylop260 thanks!

sbidoul committed:
[WIP] mis_builder refactoring: restore drilldown with support for account details

sbidoul committed:
[IMP] mis_builder: add value rendering tests, and tune the code

sbidoul committed:
[IMP] mis_builder: comparison rendering tests and tweaks

sbidoul committed:
[FIX] mis_builder: another attempt to silence a spurious pylint-odoo error

sbidoul committed:
[IMP] mis_builder: number format are now part of styles

Plus a default style at the report level.
Plus correct number rendering for comparisons in Excel export.

sbidoul committed:
[FIX] mis_builder: dependencies

sbidoul committed:
[FIX] mis_builder: tweak xlsx number format

sbidoul committed:
[IMP] mis_builder: cosmetics

sbidoul committed:
[FIX] mis_builder: multi-kpi comparison column ordering

sbidoul committed:
[IMP] mis_builder: move comparison code to the style class

Ultimately KpiMatrix should not have to know about kpi at all, it should
become a kind of spreadsheet object that can render values in row/cols/subcols
with styles. At this stage, the qweb and pdf reports, as well as the as_dict()
method of the matrix already work without knowning anything about kpis.

sbidoul committed:
[IMP] mis_builder: tweak report instance view

sbidoul committed:
[IMP] mis_builder: do not talk about periods in KpiMatrix, only columns

sbidoul committed:
[IMP] mis_builder: remove unneeded div in mis widget

sbidoul committed:
[WIP] mis_builder: add TODO related to sub-kpi naming in SimpleArray

sbidoul committed:
[IMP] mis_builder: default values for colors

sbidoul committed:
[IMP] mis_builder: restore style expressions

The addition is local and propagates automatically to the the
widget, xls and pdf exports \o/

sbidoul committed:
[FIX] mis_builder: fix sum aggregation of non-stored field in non-accounting queries

fixes #178

sbidoul committed:
[IMP] mis_builder: remove _ from some method names

because they are part of the public API (though not usable over rpc)

sbidoul committed:
[IMP] mis_builder: changelog and bump to version 2.0.0

sbidoul committed:
[FIX] mis_builder: avoid calling compute twice in widget

sbidoul commented:
@adrienpeiffer can you have a look at https://github.com/OCA/account-financial-reporting/pull/189/commits/4dd0aec548b04d80024fcfd7b244b41935ebe0c8 and check if mis_builder.js could be cleaned up of some unneeded code, now that the hacks for displaying the widget first are not necessary anymore?

sbidoul committed:
[IMP] mis_builder: simplify widget template

Plus more consistent vocabulary for row/col labels

sbidoul committed:
[IMP] mis_builder: be more consistent in naming row/column labels

- label: the main row/column label
- description: a more detailed description of the row/column,
  currently displayed as a second line in the header cell
- comment: a even more detailed comment,
  currently displayed as a tooltip in the interactive widget only

sbidoul committed:
[IMP] mis_builder: hide button box in edit mode on the report instance settings form

sbidoul committed:
[FIX] mis_builder: typos in changelog

sbidoul committed:
[FIX] mis_builder: fix test

sbidoul committed:
[IMP] mis_builder: better way to disable eval-used lint check

sbidoul committed:
[IMP] mis_builder: add a wide-coverage integration test

This reveals a couple of bugs!

sbidoul committed:
[FIX] mis_builder: handle corner case errors raised by the test suite

Plus, handle even more corner cases related to NameError handling.

sbidoul committed:
[IMP] mis_builder: attempt to silence spurious pylint error

moylop260 commented:
Could you fix the following lint checks https://travis-ci.org/OCA/account-financial-reporting/jobs/130640787#L465-L465?

sbidoul committed:
[IMP] mis_builder: more tests and fixes wrt DataError in results

sbidoul commented:
@moylop260 

> Could you fix the following lint checks https://travis-ci.org/OCA/account-financial-reporting/jobs/130640787#L465-L465?

I can't reproduce these errors on my machine and the code looks well indented to me.

``` bash
$ flake8 --version
2.5.4 (pep8: 1.7.0, mccabe: 0.2.1, pyflakes: 1.1.0) CPython 3.5.1+ on Linux
```

Do you know if they come from a specific flake8 version?

moylop260 commented:
You are right!
Maybe we will need freeze the old flake8 version [MQT/travis_install_nightly](https://github.com/OCA/maintainer-quality-tools/blob/b8be85470a1e56687f535931cdb1ab01e6647a42/travis/travis_install_nightly#L6)

Commit:
[IMP] mis_builder: Cleanup JS widget.

adrienpeiffer commented:
@sbidoul I just did a little cleaning in the widget but nothing major

moylop260 commented:
odoo recommend us [don't use api.one](https://github.com/odoo/odoo/pull/8527)
Should we remove them in this PR?

Could you review the following messages?

``` bash
mis_builder/README.rst:70 [E7901(rst-syntax-error), ]  Unexpected indentation.
mis_builder/README.rst:71 [E7901(rst-syntax-error), ] Block quote ends without a blank line; unexpected unindent.
mis_builder/security/ir.model.access.csv [W7906(duplicate-id-csv), ] Duplicate id "access_mis_report_kpi_expression"
mis_builder/security/ir.model.access.csv [W7906(duplicate-id-csv), ] Duplicate id "access_mis_report_kpi_expression"
mis_builder/security/ir.model.access.csv [W7906(duplicate-id-csv), ] Duplicate id "access_mis_report_style" in ir.model.access.csv file

mis_builder/__init__.py:1: [W7908(missing-newline-extrafiles), ] wizard/mis_builder_dashboard.xml Missing newline
mis_builder/__init__.py:1: [W7908(missing-newline-extrafiles), ] tests/mis.report.csv Missing newline
mis_builder/__init__.py:1: [W7908(missing-newline-extrafiles), ] tests/mis.report.instance.csv Missing newline
mis_builder/__init__.py:1: [W7908(missing-newline-extrafiles), ] tests/mis.report.query.csv Missing newline
mis_builder/__init__.py:1: [W7910(wrong-tabs-instead-of-spaces), ] static/src/xml/mis_widget.xml:28 Use wrong tabs indentation instead of four spaces
mis_builder/__init__.py:1: [W7910(wrong-tabs-instead-of-spaces), ] static/src/xml/mis_widget.xml:50 Use wrong tabs indentation instead of four spaces
mis_builder/models/mis_report.py:532: [C8107(translation-required), MisReportKpi._inverse_expression] String parameter of raise "UserError" requires translation. Use _('Can not update a multi kpi from the kpi line')
mis_builder/models/mis_report.py:976: [C8107(translation-required), MisReport.declare_and_compute_period] String parameter of raise "UserError" requires translation. Use _("Probably not your fault... but I'm really curious to know how you managed to raise this error so I can handle one more corner case!")
```

We have a new beta check for js-lint
Could you give us feedback of fails detected?

``` bash
mis_builder/__init__.py:1: [W7903(javascript-lint), ] static/src/js/mis_builder.js:25:24: Missing semicolon.
mis_builder/__init__.py:1: [W7903(javascript-lint), ] static/src/js/mis_builder.js:26:69: Missing semicolon.
mis_builder/__init__.py:1: [W7903(javascript-lint), ] static/src/js/mis_builder.js:35:69: Missing semicolon.
mis_builder/__init__.py:1: [W7903(javascript-lint), ] static/src/js/mis_builder.js:37:45: ['no_destroy'] is better written in dot notation.
mis_builder/__init__.py:1: [W7903(javascript-lint), ] static/src/js/mis_builder.js:43:25: Missing semicolon.
mis_builder/__init__.py:1: [W7903(javascript-lint), ] static/src/js/mis_builder.js:45:20: ['active_ids'] is better written in dot notation.
mis_builder/__init__.py:1: [W7903(javascript-lint), ] static/src/js/mis_builder.js:47:23: Missing semicolon.
mis_builder/__init__.py:1: [W7903(javascript-lint), ] static/src/js/mis_builder.js:51:24: Missing semicolon.
mis_builder/__init__.py:1: [W7903(javascript-lint), ] static/src/js/mis_builder.js:52:94: Missing semicolon.
mis_builder/__init__.py:1: [W7903(javascript-lint), ] static/src/js/mis_builder.js:62:24: Missing semicolon.
mis_builder/__init__.py:1: [W7903(javascript-lint), ] static/src/js/mis_builder.js:63:94: Missing semicolon.
mis_builder/__init__.py:1: [W7903(javascript-lint), ] static/src/js/mis_builder.js:73:24: Missing semicolon.
mis_builder/__init__.py:1: [W7903(javascript-lint), ] static/src/js/mis_builder.js:74:94: Missing semicolon.
mis_builder/__init__.py:1: [W7903(javascript-lint), ] static/src/js/mis_builder.js:84:24: Missing semicolon.
mis_builder/__init__.py:1: [W7903(javascript-lint), ] static/src/js/mis_builder.js:85:94: Missing semicolon.
mis_builder/__init__.py:1: [W7903(javascript-lint), ] static/src/js/mis_builder.js:135:24: Missing semicolon.
mis_builder/__init__.py:1: [W7903(javascript-lint), ] static/src/js/mis_builder.js:137:61: Missing semicolon.
mis_builder/__init__.py:1: [W7903(javascript-lint), ] static/src/js/mis_builder.js:138:25: ['no_destroy'] is better written in dot notation.
```

sbidoul committed:
[IMP] mis_builder: lint

sbidoul commented:
@moylop260 great work you are doing on this lint stuff!

Can you look at this crash in pylint-odoo ? https://travis-ci.org/OCA/account-financial-reporting/jobs/130776237#L508

moylop260 commented:
@sbidoul thank you
Yes the fix is: https://github.com/OCA/pylint-odoo/pull/30

moylop260 commented:
@sbidoul fixed, now you have the last messages [here](https://travis-ci.org/OCA/account-financial-reporting/jobs/130776237#L466-L486)
I detected a valid case of `copy` method without a `super`
`mis_builder/models/mis_report_style.py:18: [W8106(method-required-super), PropertyDict.copy] Missing`super`call in "copy" method.`
For this case you can use: `pylint: disable=method-required-super,other_check...`

sbidoul committed:
[IMP] mis_builder: more lint

sbidoul committed:
[IMP] mis_builder: remove deprecated api.one

sbidoul commented:
@moylop260 I'm still lost with the hanging indent errors.

I installed the latest flake8 locally in a fresh virtualenv and run it without errors. Do you have a suggestion?

ThomasBinsfeld commented:
@sbidoul Launch flake8 with --ignore ("flake8 --ignore E128 <FileName>" for example) and it will show the under/over-indented lines

sbidoul committed:
[IMP] mis_builder: more lint

sbidoul commented:
@ThomasBinsfeld thanks! I would not have thought ignoring an error would raise others.

sbidoul commented:
Ok, so now I consider the code is much cleaner and maintainable with a proper separation of concerns:
- mis_report.py: the core module providing the report templates and computation logic relying on a bunch of lower level features (aep, accounting_none, simple_array, mis_report_style)
- mis_report_instance.py: providing instanciation of report templates for given time periods
- the various renderings: widget, qweb-pdf, xslx

There is full feature parity across the 3 renderings.

For more information on new features and fixes compared to the v8 version, see the changelog: https://github.com/oca-sorrento/account-financial-reporting/blob/022b8a8998d7f38fb40f0e82185211b2f044ab62/mis_builder/CHANGES.rst

The test coverage is reasonably complete (78%, with the AccoutingExpressionProcessor fully tested except some error cases).

I still have plenty of ideas for improvements and features, but I'll stop here for now!

Happy reviewing!

moylop260 commented:
Should we fix the duplicated csv id?

```
mis_builder/__init__.py:1: [W7906(duplicate-id-csv), ] Duplicate id "access_mis_report_kpi_expression" in ir.model.access.csv file
mis_builder/__init__.py:1: [W7906(duplicate-id-csv), ] Duplicate id "access_mis_report_subkpi" in ir.model.access.csv file
mis_builder/__init__.py:1: [W7906(duplicate-id-csv), ] Duplicate id "access_mis_report_style" in ir.model.access.csv file
```

sbidoul commented:
@moylop260 perhaps, perhaps. I don't know (yet) what this warning means.

moylop260 commented:
Help me to use a clearer message for the following case:
![csv](https://cloud.githubusercontent.com/assets/6644187/15325979/dd6d73fe-1c11-11e6-8789-31527169e935.png)

[duplicate-id-csv](https://github.com/OCA/account-financial-reporting/blob/022b8a8998d7f38fb40f0e82185211b2f044ab62/mis_builder/security/ir.model.access.csv#L12-L17)

moylop260 commented:
If you have a proposal for this message you could comment here: https://github.com/OCA/pylint-odoo/pull/32

sbidoul committed:
[FIX] mis_builder: duplicate record ids in ACLs

sbidoul commented:
@moylop260 the message is just fine. Useful check, thanks! Fixed.

moylop260 commented:
Thanks @sbidoul 
👍

lepistone commented on mis_builder/datas/ir_cron.xml:
```diff
@@ -0,0 +1,18 @@
+    <data noupdate="1">
+
+        <record id="ir_cron_crm_action" model="ir.cron">
+            <field name="name">Vaccum temporary report</field>
```
Vacuum

lepistone commented on mis_builder/datas/ir_cron.xml:
```diff
@@ -0,0 +1,18 @@
+<openerp>
+    <data noupdate="1">
+
+        <record id="ir_cron_crm_action" model="ir.cron">
```
better XMLID?

lepistone commented on mis_builder/models/aep.py:
```diff
@@ -368,17 +294,146 @@ def f(mo):
+            # as it does not make sense to distinguish 0 from "no data"
+            if v is not AccountingNone and \
+                    mode in (self.MODE_INITIAL, self.MODE_UNALLOCATED) and \
+                    float_is_zero(v, precision_rounding=2):
```
Is it correct to hardcode `2` here?

sbidoul replied:
Well, ideally it should come from the kpi style which itself defaults to the report style. But bringing that value down here is a lot of code for very little benefit (this only serves to distinguish 0 from null in the initial balance).

I think I'll change 2 to 4 so be on the safe side and we'll see if it creates problems in practice (I bet not).

elicoidal replied:
Is there any reason not to take the decimal precision set up for accounting?

sbidoul replied:
@elicoidal you are right! in this area of mis_builder, we rely on accounting precision, not the precision set in the styles. So I changed to use the decimal places of the company currency, and adapted the AEP api to pass the company in the constructor instead of moving it around in each method.

elicoidal replied:
my very little piece for a huge job that you have produced so far!

sbidoul replied:
Thanks. As we say every little bit helps. And I'm glad to have had new contributors in Sorrento.

lepistone commented on mis_builder/models/aep.py:
```diff
@@ -368,17 +294,146 @@ def f(mo):
+    @classmethod
+    def get_balances_variation(cls, company, date_from, date_to,
+                               target_move='posted'):
+        """ A convenience method to obtain the variantion of the
```
variation

sbidoul committed:
[FIX] mis_builder: better xmlid and name for vacuum cron

sbidoul committed:
[FIX] mis_builder: typo in docstring

sbidoul committed:
[IMP] mis_builder: change precision rounding from 2 to 4 to distinguish 0 from null in initial balances

This should be slightly on the safer side. Ideally, this rounding precision
should come from the kpi style (which defaults to the report style), but
that would be a lot of code for little benefits.

sbidoul committed:
[IMP] mis_builder: use company currency decimal place in deciding if initial balances are null or 0

lepistone commented on mis_builder/models/simple_array.py:
```diff
@@ -0,0 +1,131 @@
+# -*- coding: utf-8 -*-
+# © 2014-2015 ACSONE SA/NV (<http://acsone.eu>)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+""" A trivial immutable array that supports basic arithmetic operations.
```
This is somewhat like `numpy.array` without pulling all of numpy, am I right?

sbidoul replied:
@lepistone it is. I kind of thought pulling numpy was not appropriate :wink: 
In the end SimpleArray does a bit more wrt exception handling (DataError).

lepistone replied:
OK!

lepistone commented:
Great work, thanks! 👍 

I think I can merge this!
